### DISCUSSION TITLE
I want to clear the LABEL of the FRAME by pressing the button.

### DISCUSSION THREAD

**Discussion by Daiki48** [fltk-rs/fltk-rs#1321] [Q&A] [answered]
I was looking for a RUST gui crate and found fltk-rs.
I am trying the [hello_button examples](https://github.com/fltk-rs/fltk-rs/blob/master/fltk/examples/hello_button.rs).

In this example, the process is to press a button to display "Hello world".
I would additionally like to add a process to clear that "hello world".

I have made it look clear by overwriting a blank LABEL on the FRAME.
I don't think this is very beautiful...

Does the button or frame have a clear method?
Or if there is a more beautiful way to write code, please let me know.

```rust
#![windows_subsystem = "windows"]

use fltk::{
    app,
    button::Button,
    frame::Frame,
    prelude::*,
    window::Window,
};

fn main() {
    let app = app::App::default();
    let mut wind = Window::default().with_size(400, 300).with_label("Hello Button!");
    let mut frame = Frame::default().with_size(200, 100).center_of(&wind);
    let mut hello_btn = Button::new(160, 210, 80, 40, "Click me!");
    let mut clear_btn = Button::new(160, 40, 80, 40, "Clear");

    wind.end();
    wind.show();

    hello_btn.set_callback({
        let mut frame = frame.clone();
        move |_| frame.set_label("Hello world")
    });

    // here
    clear_btn.set_callback({
        move |_| frame.set_label("")
    });

    app.run().unwrap();
}
```

P.S.
I am a beginner in programming, I have hardly touched any programming language except rust.
Of course, rust language has been writing for several months now.

**Comment by MoAlyousef** [comment, marked as answer]
FLTK provides the set_label method, no clear_label. You can implement your own if you want:
```rust
trait ClearLabel<W>
where
    W: WidgetExt,
{
    fn clear_label(&mut self);
}

impl<W> ClearLabel<W> for W
where
    W: WidgetExt,
{
    fn clear_label(&mut self) {
        self.set_label("");
    }
}

fn main() {
    let mut frame = Frame::new(/*params*/);
    frame.clear_label();
}
```

You can also check the [flemish](https://github.com/fltk-rs/flemish) crate, which builds an elmish architecture on top of fltk-rs, so you don't have to use methods like set_label():
```rust
use flemish::{
    button::Button, color_themes, frame::Frame, group::Flex, prelude::*, OnEvent, Sandbox, Settings,
};

pub fn main() {
    App::new().run(Settings {
        size: (300, 100),
        resizable: true,
        ..Default::default()
    })
}

#[derive(Default)]
struct App{
    value: String,
}

#[derive(Debug, Clone, Copy)]
enum Message {
    Show,
    Clear,
}

impl Sandbox for App{
    type Message = Message;

    fn new() -> Self {
        Self::default()
    }

    fn title(&self) -> String {
        String::from("Hello Button")
    }

    fn update(&mut self, message: Message) {
        match message {
            Message::Show=> {
                self.value = "Hello world".to_string();
            }
            Message::Clear=> {
                self.value.clear();
            }
        }
    }

    fn view(&mut self) {
        let col = Flex::default_fill().column();
        Button::default()
            .with_label("Click me")
            .on_event(Message::Show);
        Frame::default().with_label(&self.value);
        Button::default()
            .with_label("Clear")
            .on_event(Message::Clear);
        col.end();
    }
}
```

**Reply by Daiki48** [reply]
> FLTK provides the set_label method, no clear_label. You can implement your own if you want:

I did not know how to make my own trait. It is a great learning experience. Thank you very much.
What is this `<W>` ?
Is this the generic described in [the where clauses of Rust By Example](https://doc.rust-lang.org/rust-by-example/generics/where.html) ?

> You can also check the [flemish](https://github.com/fltk-rs/flemish) crate, which builds an elmish architecture on top of fltk-rs, so you don't have to use methods like set_label():

I had never heard of this crate before. Thanks for telling me about it.
I know iced, in a crust gui crate.
I studied elm architecture at iced.
I will also try to write in flemish.

**Reply by MoAlyousef** [reply]
> What is this <W> ?
Is this the generic described in [the where clauses of Rust By Example](https://doc.rust-lang.org/rust-by-example/generics/where.html) ?

Yes exactly. The `where` clause adds a trait bound to objects implementing WidgetExt.
This allows all widgets which implement WidgetExt to benefit from the new trait.

**Reply by Daiki48** [reply]
Thank you kindly for your kind words.
I will try to implement the clearing process myself using the code you wrote. Of course, I will also use the crate you introduced.
